File: experiment/websites/c/shellsort/shellsort.c
/* based on: http://rosettacode.org/wiki/Sorting_algorithms/Bubble_sort#C (2021-12-13) */

#include <emscripten/emscripten.h>

EMSCRIPTEN_KEEPALIVE
void shellsort (int *a, int n) {
  int h, i, j, t;
  for (h = n; h /= 2;) {
    for (i = h; i < n; i++) {
      t = a[i];
      for (j = i; j >= h && t < a[j - h]; j -= h) {
        a[j] = a[j - h];
      }
      a[j] = t;
    }
  }
}

File: experiment/websites/js/shellsort/shellsort.js
/* based on: https://rosettacode.org/wiki/Sorting_algorithms/Shell_sort#JavaScript (2021-12-23) */

/**
 * @param {number[]} a
 * @returns {number[]}
 */
function shellsort (a) {
  for (var h = a.length; h > 0; h = parseInt(h / 2)) {
    for (var i = h; i < a.length; i++) {
      var k = a[i];
      for (var j = i; j >= h && k < a[j - h]; j -= h)
        a[j] = a[j - h];
      a[j] = k;
    }
  }
  return a;
}
